### PR TITLE
Expose Ampache Token Generation API

### PR DESCRIPTION
**Pull request by gnarula**
This would help integrate the music plugin with clients that wish to use the Ampache API.

For more information, please refer to the comments at the following [doc](https://docs.google.com/document/d/189ESSu-E4kP6sXru3lxTohzXF7c8-BTDwosvtv951eI/edit?disco=AAAAAkgsNi0#heading=h.o5puy7ax05dt)

### CODE REVIEW

File: controller/ampachecontroller.php [hunk 1]
```diff
@@ -501,4 +502,26 @@ protected function get_cover() {
...
 		}
 	}
+
+	/**
+	 * this is not a part of ampache specification
+	 * This is an exposed API endpoint.
+	 * Client uses settings#addUserKey instead
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 * @CORS
```
Comment (MorrisJobke) [inline controller/ampachecontroller.php]: I would propose to not expose another API endpoint, but simply add the annotations to the existing one in https://github.com/owncloud/music/blob/acfe12afcd05447fa0dfc698b5ff5c7f930ef552/controller/settingcontroller.php#L73

Then you also don't need to add any route ;)

Comment (gnarula) [inline controller/ampachecontroller.php]: I tried to do the same initially but turns out an action annotated with `@CORS` is not allowed to be used in conjunction with session authentication https://github.com/owncloud/core/blob/master/lib/private/appframework/middleware/security/corsmiddleware.php#L82

Comment (MorrisJobke) [inline controller/ampachecontroller.php]: > I tried to do the same initially but turns out an action annotated with @CORS is not allowed to be used in conjunction with session authentication

But we use the OCS API for example also in the core sharing dialog. @LukasReschke Do you have an idea how to accomplish this?

Comment (MorrisJobke) [inline controller/ampachecontroller.php]: > Do you have an idea how to accomplish this?

Without code duplication.

Workaround: Create a second route in the `SettingController` and call in the method code just the original method to not copy the code but only change the authentication.

**Comment by MorrisJobke** [comment]
@LukasReschke Could I get your opinion on this? :)

**Comment by MorrisJobke** [comment]
I'm just thinking about this whole setUserKey method. I guess it would be better to have a "generateNewUserKey" and send it back. Otherwise this could lead to bad and weak passwords.

**Comment by gnarula** [comment]
> I guess it would be better to have a "generateNewUserKey" and send it back

Should the length of the desired password be an input in that case? Maybe we can set a minimum length (say 8?) too.

Also, is it safe to assume [openssl_random_pseudo_bytes](http://php.net/openssl_random_pseudo_bytes) would be available for generating cryptographically random data?

**Comment by gnarula** [comment]
Updated the PR. Implemented generateUserKey() in SettingsController

**Comment by MorrisJobke** [comment]
Looks really nice 👍 

Could you add a comment about this API to the readme: https://github.com/owncloud/music#authentication That would be really nice

**Comment by MorrisJobke** [comment]
Tested and works ... sorry for the long delay 👍

**Comment by MorrisJobke** [comment]
I just noticed some minor stuff. I will open an additional PR for this :)

**Pull request merged**